app/renderer/reducers/Session.js

```diff
@@ -206,6 +219,79 @@ export default function session (state = INITIAL_STATE, action) {
         runningAppiumSessions: action.sessions || [],
       };
 
+    case ENABLE_DESIRED_CAPS_EDITOR:
+      const {caps} = state;
+      let rawCaps = {};
+      for (let {name, value} of caps) {
+        rawCaps[name] = value;
+      }
+
+      return {
+        ...state,
+        isEditingDesiredCaps: true,
+        rawDesiredCaps: formatJSON.plain(rawCaps),
+        isValidCapsJson: true,
+        isValidatingCapsJson: false, // Don't start validating JSON until the user has attempted to save the JSON
+      };
+
+    case ABORT_DESIRED_CAPS_EDITOR:
+      return {
+        ...state,
+        isEditingDesiredCaps: false,
+        rawDesiredCaps: undefined,
+      };
+
+    case SAVE_RAW_DESIRED_CAPS:
+      const {rawDesiredCaps} = state;
+      try {
+        const newCaps = JSON.parse(rawDesiredCaps);
+
+        // Transform the current caps array to an object
+        let {caps:capsArray} = state;
+        let caps = {};
+        for (let {type, name, value} of capsArray) {
+          caps[name] = {type, value};
+        }
+
+        // Translate the caps JSON to array format
+        let newCapsArray = toPairs(newCaps).map(([name, value]) => ({
+          type: (caps[name] && caps[name].type === 'file' && typeof(value) === 'string') ? 'file' : getCapType(value),  // If we already have that cap and it's file type, keep the type the same
```

Comment:
technically `typeof` is an operator, not a function, and i think we should standardize on not using parens. `typeof value`. (did this somewhere above too)